### PR TITLE
Adding ryan's recursion solution for discussion

### PR DESCRIPTION
**Pull request by jimtla**

### CODE REVIEW

File: ryan-recursion.coffee [hunk 1]
```diff
@@ -0,0 +1,82 @@
+check = (cond, msg) -> cond || throw new Error msg
```
Comment (jimtla) [inline ryan-recursion.coffee]:
```
check = (cond, msg) -> 
  throw new Error msg unless cond
```

or 

```
check = (cond, msg) -> 
  unless cond
    throw new Error msg
```

I'm slightly morally opposed to the use of boolean operators for control flow.

File: ryan-recursion.coffee [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+checkEq = (a, b, msg) -> check b is a, "#{msg ? "-"}: #{b} isnt #{a}"
+checkArrayEq = (a, b, msg) -> check "#{b}" is "#{a}", "#{msg ? "-"}: #{b} isnt #{a} (!!! HACKY !!!)"
+
+buildGraph = (cliques) ->
+  map = {}
+  map[n] = {} for n in clique for clique in cliques
+  map[n][m] = true for m in c when m isnt n for n in c for c in cliques
+  return (vertex) -> (k for k, _ of map[vertex] ? {})
```
Comment (jimtla) [inline ryan-recursion.coffee]: unfortunately `_` is a super useful library, making it a bad throwaway variable. In this case you can just drop the second parameter

File: ryan-recursion.coffee [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+shortestPathAssumingReachable = (start, end, graph) ->
+  level = []
+  newLevel = () ->
+    oldLevel = level
+    level = []
+    return oldLevel
+  tree = {}
+  seen = (n) -> tree[n] isnt undefined
```
Comment (jimtla) [inline ryan-recursion.coffee]: `tree[n]?` is a nicer way of writing `tree[n] isn't undefined`

File: ryan-recursion.coffee [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  see n, tree[parent] for n in graph parent for parent in newLevel() until seen end
+  return tree[end]
+
+(->
+  graph = buildGraph [[0, 1], [0, 2], [1, 3, 4], [2, 3]]
+  checkArrayEq [1, 2], graph 0
+  checkArrayEq [0, 3, 4], graph 1
+  checkArrayEq [], graph 5
+  checkArrayEq [0, 1, 4], shortestPathAssumingReachable 0, 4, graph
+)()
```
Comment (jimtla) [inline ryan-recursion.coffee]: as mentioned, `do` does this

File: ryan-recursion.coffee [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    [[T, 7], [T, 9], [A, 10], [T, 12]]
+    [[T, 8], [B, 2]]
+    [[T, 10], [T, 13], [A, 13]]
+    [[A, 7], [B, 15]]
+    [[A, 8], [C, 12]]
+    [[A, 9], [A, 15]]
+    [[B, 10], [C, 3]]
+    [[B, 13], [C, 14]]
+    [[C, 6], [C, 7]]
+  ]
```
Comment (jimtla) [inline ryan-recursion.coffee:64]: This is a much better way to write the graph than alexey and I had...

File: ryan-recursion.coffee (hunk 1 above)
Comment (ryandm) [inline ryan-recursion.coffee]: ended up changing it to the first of those in my local copy :)

File: ryan-recursion.coffee (hunk 2 above)
Comment (ryandm) [inline ryan-recursion.coffee]: like

File: ryan-recursion.coffee (hunk 3 above)
Comment (ryandm) [inline ryan-recursion.coffee]: like (should've known that!)

File: ryan-recursion.coffee (hunk 4 above)
Comment (ryandm) [inline ryan-recursion.coffee]: sigh. I told Lu about this, all excited that I learned something new

before I even finished she was like, "wait, doesn't 'do' do that?"

File: ryan-recursion.coffee (hunk 5 above)
Comment (ryandm) [inline ryan-recursion.coffee:64]: I try :)

Comment (ryandm) [inline ryan-recursion.coffee:64]: generally I've found that modeling data as close as possible to its actual form is a good starting point. Here this strategy happened to work well

**Comment by ryandm** [comment]
thanks for the feedback!

time for me to read some good coffeescript example aka your pull request :)

File: ryan-recursion.coffee (hunk 1 above)
Comment (jimtla) [inline ryan-recursion.coffee]: Nice! I'm still not sure how I feel about conditions on the right side of expressions, but there are definitely situations where it makes for pretty code.

File: ryan-recursion.coffee (hunk 4 above)
Comment (jimtla) [inline ryan-recursion.coffee]: Sad. I hate it when my brilliant revelations are obvious to other people :)

File: ryan-recursion.coffee (hunk 5 above)
Comment (jimtla) [inline ryan-recursion.coffee:64]: Solid - I will try to keep that in mind next time.

**Comment by jimtla** [comment]
Any time - read critically on mine, there's probably a few ugly things and poor decisions in there.

File: ryan-recursion.coffee (hunk 1 above)
Comment (ryandm) [inline ryan-recursion.coffee]: Agreed. It can be hard to read. In this case it made sense to me.

**Comment by ryandm** [comment]
Going for a run; then I will do. The other day I discovered that you have a mapping of key code to character that doesn't agree with my programmer Dvorak keyboard layout but no other insights so far

**Pull request closed without merging**